src/root/POSRootChainManager.ts

```diff
@@ -0,0 +1,191 @@
+
+    const _options = await this._fillOptions(options, txObject, this.web3Client.getParentWeb3())
+
+    if (options.encodeAbi) {
+      _options.data = txObject.encodeABI()
+      _options.to = this.posRootChainManager.options.address
+      return _options
+    }
+    return this.web3Client.send(txObject, _options)
+  }
+
+  private async _buildPayloadForExit(burnTxHash) {
+    // check checkpoint
+    const lastChildBlock = await this.rootChain.getLastChildBlock()
+    const burnTx = await this.web3Client.getMaticWeb3().eth.getTransaction(burnTxHash)
+
+    const receipt = await this.web3Client.getMaticWeb3().eth.getTransactionReceipt(burnTxHash)
+
+    const block: any = await this.web3Client
+      .getMaticWeb3()
+      .eth.getBlock(burnTx.blockNumber, true /* returnTransactionObjects */)
+
+    logger.info({ 'burnTx.blockNumber': burnTx.blockNumber, lastCheckPointedBlockNumber: lastChildBlock })
+    assert.ok(
+      new BN(lastChildBlock).gte(new BN(burnTx.blockNumber)),
+      'Burn transaction has not been checkpointed as yet'
+    )
+    const headerBlockNumber = await this.rootChain.findHeaderBlockNumber(burnTx.blockNumber)
+    const headerBlock = await this.web3Client.call(
+      this.rootChain.getRawContract().methods.headerBlocks(this.encode(headerBlockNumber))
+    )
+    logger.info({ headerBlockNumber: headerBlockNumber.toString(), headerBlock })
+
+    // build block proof
+    const blockProof = await Proofs.buildBlockProof(
+      this.web3Client.getMaticWeb3(),
+      headerBlock.start,
+      headerBlock.end,
+      burnTx.blockNumber
+    )
+
+    const receiptProof = await Proofs.getReceiptProof(receipt, block, this.web3Client.getMaticWeb3())
+
+    const logIndex = receipt.logs.findIndex(log => log.topics[0].toLowerCase() == this.TRANSFER_EVENT_SIG)
+    assert.ok(logIndex > -1, 'Burn log not found in receipt')
+
+    return this._encodePayload(
+      headerBlockNumber,
+      blockProof,
+      burnTx.blockNumber,
+      block.timestamp,
+      Buffer.from(block.transactionsRoot.slice(2), 'hex'),
+      Buffer.from(block.receiptsRoot.slice(2), 'hex'),
+      Proofs.getReceiptBytes(receipt), // rlp encoded
+      receiptProof.parentNodes,
+      receiptProof.path,
+      logIndex,
+      burnTxHash
+    )
+  }
+
+  private _encodePayload(
+    headerNumber,
+    buildBlockProof,
+    blockNumber,
+    timestamp,
+    transactionsRoot,
+    receiptsRoot,
+    receipt,
+    receiptParentNodes,
+    path,
+    logIndex,
+    burnTxHash
+  ) {
+    return ethUtils.bufferToHex(
+      ethUtils.rlp.encode([
+        headerNumber,
+        buildBlockProof,
+        blockNumber,
+        timestamp,
+        ethUtils.bufferToHex(transactionsRoot),
+        ethUtils.bufferToHex(receiptsRoot),
+        ethUtils.bufferToHex(receipt),
+        ethUtils.bufferToHex(ethUtils.rlp.encode(receiptParentNodes)),
+        ethUtils.bufferToHex(ethUtils.rlp.encode(path)),
+        logIndex,
+        burnTxHash,
+      ])
+    )
+  }
+}
```

Comment:
This chunk is duplicated (except an extra burnTxHash field) WithdrawManager.ts. Placing them in common might help.

Reply:
Moved this to `ExitManager` class.